File: src/main/go/g1101_1200/s1143_longest_common_subsequence/solution.go
package s1143_longest_common_subsequence

// #Medium #Top_100_Liked_Questions #String #Dynamic_Programming #LeetCode_75_DP/Multidimensional
// #Algorithm_II_Day_17_Dynamic_Programming #Dynamic_Programming_I_Day_19
// #Udemy_Dynamic_Programming #Big_O_Time_O(n*m)_Space_O(n*m)
// #2025_05_11_Time_11_ms_(84.94%)_Space_20.05_MB_(31.35%)

func longestCommonSubsequence(text1 string, text2 string) int {
	n := len(text1)
	m := len(text2)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if text1[i-1] == text2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}
	return dp[n][m]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/main/swift/g1101_1200/s1143_longest_common_subsequence/Solution.swift
// #Medium #Top_100_Liked_Questions #String #Dynamic_Programming #LeetCode_75_DP/Multidimensional
// #Algorithm_II_Day_17_Dynamic_Programming #Dynamic_Programming_I_Day_19
// #Udemy_Dynamic_Programming #Big_O_Time_O(n*m)_Space_O(n*m)
// #2024_07_03_Time_12_ms_(98.60%)_Space_16.1_MB_(98.60%)

class Solution {
    func longestCommonSubsequence(_ text1: String, _ text2: String) -> Int {
        let text1 = Array(text1), text2 = Array(text2)
        let n = text1.count 
        let m = text2.count
        var dp = [Int](repeating: 0, count: m + 1)
        for i in 0..<n {
            var curRow = [Int](repeating: 0, count: m + 1)
            for j in 0..<m {
                if text1[i] == text2[j] {
                    curRow[j + 1] = 1 + dp[j]
                } else {
                    curRow[j + 1] = max(dp[j + 1], curRow[j])
                }
            }
            dp = curRow
        }
        return dp[m]
    }
}
